sys/riscv/pmap.c

```diff
@@ -1,48 +1,122 @@
+ * | user write   | 1 | 1 | 1    | * | 1 | 1 | 1 |
+ * +--------------+---+---+------+---+---+---+---+
+ * | user exec    | * | 1 | 1    | 1 | * | * | 1 |
+ * +--------------+---+---+------+---+---+---+---+
+ * | kernel read  | * | 1 | 0    | * | * | 1 | 1 |
+ * +--------------+---+---+------+---+---+---+---+
+ * | kernel write | 1 | 1 | 0    | * | 1 | 1 | 1 |
+ * +--------------+---+---+------+---+---+---+---+
+ * | kernel exec  | * | 1 | 0    | 1 | * | * | 1 |
+ * +--------------+---+---+------+---+---+---+---+
+ *
+ * The dirty (D) and accessed (A) bits may be managed automaticaly
+ * by hardware. In such case, setting of these bits is imperceptible from the
+ * perspective of the software. To be compliant with the other ports,
+ * we assume these bits to be unsupported and emulate them in software.
+ */
+
+static const pte_t pte_common = PTE_A | PTE_V;
+
+static const pte_t vm_prot_map[] = {
+  [VM_PROT_READ] = PTE_SW_READ | PTE_R | pte_common,
+  [VM_PROT_WRITE] = PTE_SW_WRITE | PTE_D | PTE_W | PTE_R | pte_common,
+  [VM_PROT_READ | VM_PROT_WRITE] =
+    PTE_SW_WRITE | PTE_SW_READ | PTE_D | PTE_W | PTE_R | pte_common,
+  [VM_PROT_EXEC] = pte_common,
+  [VM_PROT_READ | VM_PROT_EXEC] = PTE_SW_READ | PTE_X | PTE_R | pte_common,
+  [VM_PROT_WRITE | VM_PROT_EXEC] =
+    PTE_SW_WRITE | PTE_D | PTE_X | PTE_W | PTE_R | pte_common,
+  [VM_PROT_READ | VM_PROT_WRITE | VM_PROT_EXEC] =
+    PTE_SW_WRITE | PTE_SW_READ | PTE_D | PTE_X | PTE_W | PTE_R | pte_common,
+};
+
+/*
+ * The list of all the user pmaps.
+ * Order of acquiring locks:
+ *  - `pmap_kernel()->mtx`
+ *  - `user_pmaps_lock`
+ */
+static MTX_DEFINE(user_pmaps_lock, 0);
+static LIST_HEAD(, pmap) user_pmaps = LIST_HEAD_INITIALIZER(user_pmaps);
+
 /*
  * Page directory.
  */
 
 pde_t pde_make(int lvl, paddr_t pa) {
-  panic("Not implemented!");
-}
-
-void broadcast_kernel_top_pde(vaddr_t va, pde_t pde) {
-  panic("Not implemented!");
+  return PA_TO_PTE(pa) | PTE_V;
 }
 
 /*
  * Page table.
  */
 
 pte_t pte_make(paddr_t pa, vm_prot_t prot, unsigned flags) {
-  panic("Not implemented!");
+  pte_t pte = PA_TO_PTE(pa) | vm_prot_map[prot];
+  bool kernel = flags & _PMAP_KERNEL;
+
+  const pte_t mask_on = kernel ? PTE_G : PTE_U;
+  pte |= mask_on;
+
+  const pte_t mask_off = kernel ? 0 : PTE_D | PTE_A | PTE_W | PTE_V;
+  pte &= ~mask_off;
+
+  /* TODO(MichalBlk): if the target board supports PMA setting,
+   * set the attributes according to cache flags passed in `flags`. */
+
+  return pte;
 }
 
 inline pte_t pte_protect(pte_t pte, vm_prot_t prot) {
-  panic("Not implemented!");
+  return (pte & ~PTE_PROT_MASK) | vm_prot_map[prot];
 }
 
 /*
  * Physical map management.
  */
 
-void pmap_md_setup(pmap_t *pmap) {
-  panic("Not implemented!");
+void pmap_md_activate(pmap_t *umap) {
+  /* Update kernel page directory entries within the pmap. */
+  pmap_t *kmap = pmap_kernel();
+  size_t halfpage = PAGESIZE / 2;
+  WITH_MTX_LOCK (&kmap->mtx) {
+    memcpy((void *)phys_to_dmap(umap->pde) + halfpage,
```

Comment:
While I suggested lazy solution for the problem, I'd like not to copy half of page directory each time we switch address spaces. Perhaps we could add generation counter for each `pmap`. Generation for a pmap would increase if somebody called `pmap_growkernel`. When we switch to another pmap and active pmap's generation counter is more recent than ours, we would do the copy. @MichalBlk @pj1031999 What do you think?

Reply:
Is it observable performance issue? We have other critical paths in kernel that are slow and inefficient. As long as nobody will run fully profiled kernel under heavy multithread load I don't really care about single function. How much instruction it is? 256 moves? I guess that mutex lock in previous line is more expensive.

Anyway, I opt for leaving it as is now, merge support for RISC-V and then focus on fixing existing bad behavior issues. Next we should invest in performance issue since our human resources are super limited ;)

Reply:
@MichalBlk accepted the challenge and implemented it ;-)